src/main/java/org/spongepowered/common/service/pagination/SpongePaginationService.java

```diff
@@ -0,0 +1,218 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.service.pagination;
+
+import static org.spongepowered.api.util.command.args.GenericArguments.integer;
+import static org.spongepowered.common.util.SpongeCommonTranslationHelper.t;
+
+import com.google.common.base.Functions;
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.MapMaker;
+import net.minecraft.entity.player.EntityPlayerMP;
+import org.spongepowered.api.service.pagination.PaginationBuilder;
+import org.spongepowered.api.service.pagination.PaginationCalculator;
+import org.spongepowered.api.service.pagination.PaginationService;
+import org.spongepowered.api.text.Text;
+import org.spongepowered.api.util.StartsWithPredicate;
+import org.spongepowered.api.util.command.CommandException;
+import org.spongepowered.api.util.command.CommandResult;
+import org.spongepowered.api.util.command.CommandSource;
+import org.spongepowered.api.util.command.args.ArgumentParseException;
+import org.spongepowered.api.util.command.args.CommandArgs;
+import org.spongepowered.api.util.command.args.CommandContext;
+import org.spongepowered.api.util.command.args.CommandElement;
+import org.spongepowered.api.util.command.spec.CommandExecutor;
+import org.spongepowered.api.util.command.spec.CommandSpec;
+import org.spongepowered.common.Sponge;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import javax.annotation.Nullable;
+
+public class SpongePaginationService implements PaginationService {
+
+    static class SourcePaginations {
+        private final Map<UUID, ActivePagination> paginations = new ConcurrentHashMap<UUID, ActivePagination>();
+        private volatile UUID lastUuid;
+
+        public ActivePagination get(UUID uuid) {
+            return this.paginations.get(uuid);
+        }
+
+        public void put(ActivePagination pagination) {
+            synchronized (this.paginations) {
+                this.paginations.put(pagination.getId(), pagination);
+                this.lastUuid = pagination.getId();
+            }
+        }
+
+        public Set<UUID> keys() {
+            return this.paginations.keySet();
+        }
+
+        public UUID getLastUuid() {
+            return this.lastUuid;
+        }
+    }
+    final ConcurrentMap<Class<? extends CommandSource>, PaginationCalculator<?>> calculators = new ConcurrentHashMap<Class<? extends
+            CommandSource>, PaginationCalculator<?>>();
+    final ConcurrentMap<CommandSource, SourcePaginations> activePaginations = new MapMaker().weakKeys().makeMap();
+
+    public SpongePaginationService() {
+        Sponge.getGame().getCommandDispatcher().register(Sponge.getPlugin(), CommandSpec.builder()
```

Comment:
Is it possible to __not__ register the command if the pagination service is not used? The API doesn't allow removing services (or changing the provider) for some stupid reason, and removing commands isn't exposed without casting.